### PR TITLE
Rolling window

### PR DESCRIPTION
**Pull request by michaelgsharp**
This PR depends on PR #4993, because that PR is not in yet this PR says it is making more changes then it is. The only new changes this pr brings in once that PR goes in is the RollingWindow featurizer. So you can only review the RollingWindow files and tests and can ignore the others. This branch will be updated when that other PR gets merged.

RollingWindow does rolling window calculations per grain (kinda like a composite unique key by combining multiple columns). It currently provides min/mean/max.

### CODE REVIEW

**Comment by michaelgsharp** [comment]
@harishsk Here is one of the draft PR's. As mentioned in my PR comment it depends on PR #4993, so I will work on fixing the comments and build errors for that one. Because that PR is still open, all the featurizers that are already in master say they are getting changed. When that PR is closed I'll update this branch, so you only need to worry about reviewing the RollingWindow files in this pr.

File: src/Microsoft.ML.Featurizers/RollingWindowFeaturizer.cs [hunk 1]
```diff
@@ -0,0 +1,1028 @@
...
+        }
+
+        internal sealed class Options: TransformInputBase
+        {
+            [Argument((ArgumentType.MultipleUnique | ArgumentType.Required), HelpText = "List of grain columns",
+                Name = "GrainColumns", ShortName = "grains", SortOrder = 0)]
+            public string[] GrainColumns;
+
+            [Argument(ArgumentType.Multiple | ArgumentType.Required, HelpText = "New column definition (optional form: name:src)",
+                Name = "Columns", ShortName = "col", SortOrder = 1)]
```
Comment (ganik) [inline src/Microsoft.ML.Featurizers/RollingWindowFeaturizer.cs:150]: >Columns [](start = 24, length = 7)

Make sure its a "Column", singular. This is a standard, see other ones like ToString or RobustScaler. Unfortunately putting plural is not handled down the line in code generation in NimbusML

File: src/Microsoft.ML.Data/DataView/DataViewConstructionUtils.cs [hunk 2]
```diff
@@ -80,6 +80,9 @@ internal static SchemaDefinition GetSchemaDefinition<TRow>(IHostEnvironment env,
                         schemaDefinitionCol.AddAnnotation(annotation.Name , info);
                     }
                 }
+                // REVIEW: we are dropping schema information when the schema is passed in.
```
Comment (michaelgsharp) [inline src/Microsoft.ML.Data/DataView/DataViewConstructionUtils.cs:83]: change review to TODO

File: src/Microsoft.ML.Featurizers/RollingWindowFeaturizer.cs [hunk 3]
```diff
@@ -0,0 +1,1028 @@
...
+        /// <param name="grainColumns">The list of grain columns</param>
+        /// <param name="outputColumn">Where to store the result of the calculation</param>
+        /// <param name="windowCalculation">The window calculation to perform</param>
+        /// <param name="horizon">The horizon value</param>
+        /// <param name="maxWindowSize">The maximum number of items in the window</param>
+        /// <param name="minWindowSize">The minimum number of items required. If there are less, double.NaN is returned.</param>
+        /// <param name="inputColumn">The source column.</param>
+        /// <returns></returns>
+        public static RollingWindowEstimator RollingWindow(this TransformsCatalog catalog, string[] grainColumns, string outputColumn, RollingWindowEstimator.RollingWindowCalculation windowCalculation,
+            UInt32 horizon, UInt32 maxWindowSize, UInt32 minWindowSize = 1, string inputColumn = null)
```
Comment (michaelgsharp) [inline src/Microsoft.ML.Featurizers/RollingWindowFeaturizer.cs:51]: Add more comments about horizon in the common.cs file.

Comment (michaelgsharp) [reply src/Microsoft.ML.Featurizers/RollingWindowFeaturizer.cs:51]: review naming for public api

File: src/Microsoft.ML.Featurizers/RollingWindowFeaturizer.cs [hunk 4]
```diff
@@ -0,0 +1,1028 @@
...
+    /// <format type="text/markdown"><![CDATA[
+    ///
+    /// ###  Estimator Characteristics
+    /// |  |  |
+    /// | -- | -- |
+    /// | Does this estimator need to look at the data to train its parameters? | No |
+    /// | Input column data type | double |
+    /// | Output column data type | vector of double. Size of vector is equal to the horizon |
+    ///
+    /// The <xref:Microsoft.ML.Transforms.RollingWindowEstimator> is a trivial estimator and doesn't need training.
```
Comment (michaelgsharp) [inline src/Microsoft.ML.Featurizers/RollingWindowFeaturizer.cs:109]: Add the additional comments here.

File: src/Microsoft.ML.Featurizers/RollingWindowFeaturizer.cs [hunk 5]
```diff
@@ -0,0 +1,1028 @@
...
+            }
+
+            internal abstract void CreateTransformerFromEstimator(IDataView input);
+            private protected abstract unsafe void CreateTransformerFromSavedDataHelper(byte* rawData, IntPtr dataSize);
+            private protected abstract bool CreateTransformerSaveDataHelper(out IntPtr buffer, out IntPtr bufferSize, out IntPtr errorHandle);
+            private protected abstract bool GetStateHelper(TransformerEstimatorSafeHandle estimator, out TrainingState trainingState, out IntPtr errorHandle);
+            private protected abstract bool OnDataCompletedHelper(TransformerEstimatorSafeHandle estimator, out IntPtr errorHandle);
+            public abstract void Dispose();
+
+            public abstract Type ReturnType();
```
Comment (michaelgsharp) [inline src/Microsoft.ML.Featurizers/RollingWindowFeaturizer.cs:479]: Add TODO. should be able to move the majority of this class into common.

File: src/Microsoft.ML.Featurizers/RollingWindowFeaturizer.cs [hunk 6]
```diff
@@ -0,0 +1,1028 @@
...
+            internal unsafe override VBuffer<double> Transform(IntPtr grainsArray, IntPtr grainsArraySize, double input)
+            {
+                var success = TransformDataNative(TransformerHandler, grainsArray, grainsArraySize, input, out IntPtr outputCols, out IntPtr outputRows, out double* output, out IntPtr errorHandle);
+                if (!success)
+                    throw new Exception(GetErrorDetailsAndFreeNativeMemory(errorHandle));
+
+                using var handler = new TransformedVectorDataSafeHandle(new IntPtr(output), outputCols, outputRows, DestroyTransformedDataNative);
+
+                var outputArray = new double[outputCols.ToInt32()];
+
```
Comment (michaelgsharp) [inline src/Microsoft.ML.Featurizers/RollingWindowFeaturizer.cs:716]: Add comment here.

**Comment by michaelgsharp** [comment]
Closing PR to open in new branch.

**Pull request closed without merging**